File: Business Rule/Assemblies/Consolidation/Consolidation/Consolidation/data management_extensibility rules/extender_sql_migrations.vb
Imports System
Imports System.Collections.Generic
Imports System.Data
Imports System.Data.Common
Imports System.Globalization
Imports System.IO
Imports System.Linq
Imports Microsoft.VisualBasic
Imports OneStream.Finance.Database
Imports OneStream.Finance.Engine
Imports OneStream.Shared.Common
Imports OneStream.Shared.Database
Imports OneStream.Shared.Engine
Imports OneStream.Shared.Wcf
Imports OneStream.Stage.Database
Imports OneStream.Stage.Engine

Namespace Workspace.__WsNamespacePrefix.__WsAssemblyName.BusinessRule.Extender.extender_sql_migrations
	Public Class MainClass
		Public Function Main(ByVal si As SessionInfo, ByVal globals As BRGlobals, ByVal api As Object, ByVal args As ExtenderArgs) As Object
			'Get assembly migrations class
			Dim migrationsClass As New Workspace.__WsNamespacePrefix.__WsAssemblyName.Migrations()
			
			Try
				Select Case args.FunctionType
					
					Case Is = ExtenderFunctionType.Unknown
						
					Case Is = ExtenderFunctionType.ExecuteDataMgmtBusinessRuleStep
						'Get function name
						Dim functionName As String = args.NameValuePairs("p_function")
						
						'Control function name
						
						#Region "Up Migration"
						
						If functionName = "UpMigration"
							Using dbConn As DbConnInfo = BRApi.Database.CreateApplicationDbConnInfo(si)
								'Execute migration queries
								For Each query In migrationsClass.GetMigrationQueries(si, "up")
									BRApi.Database.ExecuteSql(dbConn, query, False)
								Next
								'Populate tables
								For Each query In migrationsClass.GetPopulationQueries(si, "up")
									BRApi.Database.ExecuteSql(dbConn, query, False)
								Next 
							End Using
							
						#End Region
						
						#Region "Create Tables"
						
						Else If functionName = "CreateTables"
							Using dbConn As DbConnInfo = BRApi.Database.CreateApplicationDbConnInfo(si)
								'Execute migration queries
								For Each query In migrationsClass.GetMigrationQueries(si, "up")
									BRApi.Database.ExecuteSql(dbConn, query, False)
								Next
							End Using
							
						#End Region
						
						#Region "Populate Tables"
						
						Else If functionName = "PopulateTables"
							Using dbConn As DbConnInfo = BRApi.Database.CreateApplicationDbConnInfo(si)
								'Populate tables
								For Each query In migrationsClass.GetPopulationQueries(si, "up")
									BRApi.Database.ExecuteSql(dbConn, query, False)
								Next 
							End Using
						
						#End Region
						
						#Region "Delete Tables"
						
						Else If functionName = "DeleteTables"
							Using dbConn As DbConnInfo = BRApi.Database.CreateApplicationDbConnInfo(si)
								'Populate tables
								Dim queries As List(Of String) = migrationsClass.GetMigrationQueries(si, "down")
								queries.Reverse()
								For Each query In queries
									BRApi.Database.ExecuteSql(dbConn, query, False)
								Next 
							End Using
						
						#End Region
						
						End If
					Case Is = ExtenderFunctionType.ExecuteExternalDimensionSource
						
				End Select

				Return Nothing
			Catch ex As Exception
				Throw ErrorHandler.LogWrite(si, New XFException(si, ex))
			End Try
		End Function
	End Class
End Namespace

File: Business Rule/Assemblies/Workforce Planning (WPP)/Admin Management Dashboard (AMD)/amd/data management_extensibility rules/extender_sql_migrations.cs
using System;
using System.Collections.Generic;
using System.Data;
using System.Data.Common;
using System.Globalization;
using System.IO;
using System.Linq;
using OneStream.Finance.Database;
using OneStream.Finance.Engine;
using OneStream.Shared.Common;
using OneStream.Shared.Database;
using OneStream.Shared.Engine;
using OneStream.Shared.Wcf;
using OneStream.Stage.Database;
using OneStream.Stage.Engine;

namespace Workspace.__WsNamespacePrefix.__WsAssemblyName.BusinessRule.Extender.extender_sql_migrations
{
	public class MainClass
	{
		public object Main(SessionInfo si, BRGlobals globals, object api, ExtenderArgs args)
		{
			//Get assembly migrations class
			var migrationsClass = new Workspace.__WsNamespacePrefix.__WsAssemblyName.Migrations();
			
			try
			{
				switch (args.FunctionType)
				{
					case ExtenderFunctionType.Unknown:
						break;
						
					case ExtenderFunctionType.ExecuteDataMgmtBusinessRuleStep:
						//Get function name
						string functionName = args.NameValuePairs["p_function"];
						
						//Control function name
						
						#region "Up Migration"
						
						if (functionName == "UpMigration")
						{
							using (DbConnInfo dbConn = BRApi.Database.CreateApplicationDbConnInfo(si))
							{
								//Execute migration queries
								foreach (var query in migrationsClass.GetMigrationQueries(si, "up"))
								{
									BRApi.Database.ExecuteSql(dbConn, query, false);
								}
								//Populate tables
								foreach (var query in migrationsClass.GetPopulationQueries(si, "up"))
								{
									BRApi.Database.ExecuteSql(dbConn, query, false);
								}
							}
						}
						
						#endregion
						
						#region "Create Tables"
						
						else if (functionName == "CreateTables")
						{
							using (DbConnInfo dbConn = BRApi.Database.CreateApplicationDbConnInfo(si))
							{
								//Execute migration queries
								foreach (var query in migrationsClass.GetMigrationQueries(si, "up"))
								{
									BRApi.Database.ExecuteSql(dbConn, query, false);
								}
							}
						}
						
						#endregion
						
						#region "Populate Tables"
						
						else if (functionName == "PopulateTables")
						{
							using (DbConnInfo dbConn = BRApi.Database.CreateApplicationDbConnInfo(si))
							{
								//Populate tables
								foreach (var query in migrationsClass.GetPopulationQueries(si, "up"))
								{
									BRApi.Database.ExecuteSql(dbConn, query, false);
								}
							}
						}
						
						#endregion
						
						#region "Delete Tables"
						
						else if (functionName == "DeleteTables")
						{
							using (DbConnInfo dbConn = BRApi.Database.CreateApplicationDbConnInfo(si))
							{
								//Populate tables
								List<string> queries = migrationsClass.GetMigrationQueries(si, "down");
								queries.Reverse();
								foreach (var query in queries)
								{
									BRApi.Database.ExecuteSql(dbConn, query, false);
								}
							}
						}
						
						#endregion
						
						break;
					case ExtenderFunctionType.ExecuteExternalDimensionSource:
						break;
				}

				return null;
			}
			catch (Exception ex)
			{
				throw ErrorHandler.LogWrite(si, new XFException(si, ex));
			}
		}
	}
}
